<!DOCTYPE html
  PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
  "DTD/xhtml1-strict.dtd">
<html>
  <head>
    <title>Class Hierarchy</title>
    <link href="apidocs.css" type="text/css" rel="stylesheet" />
  </head>
  <body>
    <h1>Class Hierarchy</h1>
    <ul>
    <li>demuxipy.lib.sequence.SequencingRead<ul><li><a name="sequence.fasta.FastaSequence"></a><a href="sequence.fasta.FastaSequence.html">sequence.fasta.FastaSequence</a> - <span>Represents fasta and fasta + qual sequences, attributes, and methods</span></li></ul></li><li><a name="sequence.fasta.FastaQualityReader"></a><a href="sequence.fasta.FastaQualityReader.html">sequence.fasta.FastaQualityReader</a> - <span>Represents an iterator over fasta+qual sequences from a file</span></li><li><a name="sequence.fasta.FastaReader"></a><a href="sequence.fasta.FastaReader.html">sequence.fasta.FastaReader</a> - <span>Represents an iterator over fasta sequences from a file</span></li><li><a name="sequence.fasta.FastaWriter"></a><a href="sequence.fasta.FastaWriter.html">sequence.fasta.FastaWriter</a> - <span>Write fasta and fasta+qual objects to a file</span></li><li><a name="sequence.fastq.FastqReader"></a><a href="sequence.fastq.FastqReader.html">sequence.fastq.FastqReader</a> - <span>Represents an iterator over fastaq sequences from a file</span></li><li><a name="sequence.fastq.FastqSummarizer"></a><a href="sequence.fastq.FastqSummarizer.html">sequence.fastq.FastqSummarizer</a> - <span>Represents a class for deriving summary data across a number of 
reads.</span></li><li><a name="sequence.fastq.FastqWriter"></a><a href="sequence.fastq.FastqWriter.html">sequence.fastq.FastqWriter</a> - <span>Write fastq objects to a file</span></li><li><a name="sequence.sequence.SequencingRead"></a><a href="sequence.sequence.SequencingRead.html">sequence.sequence.SequencingRead</a> - <span>A generic class encapuslating DNA sequence related methods and 
attributes</span><ul><li><a name="sequence.fastq.FastqSequencingRead"></a><a href="sequence.fastq.FastqSequencingRead.html">sequence.fastq.FastqSequencingRead</a> - <span>Represents fastq sequences, attributes, and methods.  Default `type` is 
using Sanger-bases qualities</span><ul><li><a name="sequence.fastq.FastqIlluminaRead"></a><a href="sequence.fastq.FastqIlluminaRead.html">sequence.fastq.FastqIlluminaRead</a> - <span>Represents valid attributed for illumina format sequence reads</span></li><li><a name="sequence.fastq.FastqSangerRead"></a><a href="sequence.fastq.FastqSangerRead.html">sequence.fastq.FastqSangerRead</a> - <span>Represents valid attributed for sanger format sequence reads</span></li><li><a name="sequence.fastq.FastqSolexaRead"></a><a href="sequence.fastq.FastqSolexaRead.html">sequence.fastq.FastqSolexaRead</a> - <span>Represents valid attributed for solexa format sequence reads</span></li></ul></li></ul></li><li>unittest.TestCase<ul><li><a name="sequence.tests.test_fasta.TestFastaQualReader"></a><a href="sequence.tests.test_fasta.TestFastaQualReader.html">sequence.tests.test_fasta.TestFastaQualReader</a> - <span class="undocumented">No class docstring; 1/3 methods documented</span></li><li><a name="sequence.tests.test_fasta.TestFastaReader"></a><a href="sequence.tests.test_fasta.TestFastaReader.html">sequence.tests.test_fasta.TestFastaReader</a> - <span class="undocumented">No class docstring; 1/3 methods documented</span></li><li><a name="sequence.tests.test_fasta.TestFastaWriter"></a><a href="sequence.tests.test_fasta.TestFastaWriter.html">sequence.tests.test_fasta.TestFastaWriter</a> - <span class="undocumented">No class docstring; 2/4 methods documented</span></li><li><a name="sequence.tests.test_fasta.TestSequenceSetQuality"></a><a href="sequence.tests.test_fasta.TestSequenceSetQuality.html">sequence.tests.test_fasta.TestSequenceSetQuality</a> - <span class="undocumented">No class docstring; 1/3 methods documented</span></li><li><a name="sequence.tests.test_fastq.TestFastqClassMethods"></a><a href="sequence.tests.test_fastq.TestFastqClassMethods.html">sequence.tests.test_fastq.TestFastqClassMethods</a> - <span>Genral test of fastq methods</span></li><li><a name="sequence.tests.test_fastq.TestFastqConversions"></a><a href="sequence.tests.test_fastq.TestFastqConversions.html">sequence.tests.test_fastq.TestFastqConversions</a> - <span>test conversions from one format (sanger) to another (solexa)</span></li><li><a name="sequence.tests.test_fastq.TestFastqErrors"></a><a href="sequence.tests.test_fastq.TestFastqErrors.html">sequence.tests.test_fastq.TestFastqErrors</a> - <span class="undocumented">No class docstring; 1/2 methods documented</span></li><li><a name="sequence.tests.test_fastq.TestFastqReader"></a><a href="sequence.tests.test_fastq.TestFastqReader.html">sequence.tests.test_fastq.TestFastqReader</a> - <span>test ability to parse fastq files</span></li><li><a name="sequence.tests.test_fastq.TestFastqSummarizer"></a><a href="sequence.tests.test_fastq.TestFastqSummarizer.html">sequence.tests.test_fastq.TestFastqSummarizer</a> - <span>test our methods to summarize fastq read data</span></li><li><a name="sequence.tests.test_fastq.TestFastqTrimmingMaskingAndSlicingMethods"></a><a href="sequence.tests.test_fastq.TestFastqTrimmingMaskingAndSlicingMethods.html">sequence.tests.test_fastq.TestFastqTrimmingMaskingAndSlicingMethods</a> - <span>test snapshotting, trimming, masking, masking and trimmming, and 
slicing</span></li><li><a name="sequence.tests.test_fastq.TestFastqWriter"></a><a href="sequence.tests.test_fastq.TestFastqWriter.html">sequence.tests.test_fastq.TestFastqWriter</a> - <span>test ability to write fastq files</span></li><li><a name="sequence.tests.test_fastq.TestNonSangerQualitySpecMethods"></a><a href="sequence.tests.test_fastq.TestNonSangerQualitySpecMethods.html">sequence.tests.test_fastq.TestNonSangerQualitySpecMethods</a> - <span>Ensure that invalid Solexa quality values fail</span></li><li><a name="sequence.tests.test_fastq.TestSangerQualitySpecMethods"></a><a href="sequence.tests.test_fastq.TestSangerQualitySpecMethods.html">sequence.tests.test_fastq.TestSangerQualitySpecMethods</a> - <span>Ensure that invalid Sanger quality values fail</span></li><li><a name="sequence.tests.test_fops.TestFopsFunctions"></a><a href="sequence.tests.test_fops.TestFopsFunctions.html">sequence.tests.test_fops.TestFopsFunctions</a> - <span class="undocumented">No class docstring; 7/10 methods documented</span></li><li><a name="sequence.tests.test_sequence.TestSequenceMethods"></a><a href="sequence.tests.test_sequence.TestSequenceMethods.html">sequence.tests.test_sequence.TestSequenceMethods</a> - <span class="undocumented">No class docstring; 11/12 methods documented</span></li><li><a name="sequence.tests.test_sequence.TestSequenceSetQuality"></a><a href="sequence.tests.test_sequence.TestSequenceSetQuality.html">sequence.tests.test_sequence.TestSequenceSetQuality</a> - <span class="undocumented">No class docstring; 1/3 methods documented</span></li><li><a name="sequence.tests.test_sequence.TestSequenceTrimmingMethods"></a><a href="sequence.tests.test_sequence.TestSequenceTrimmingMethods.html">sequence.tests.test_sequence.TestSequenceTrimmingMethods</a> - <span class="undocumented">No class docstring; 12/14 methods documented</span></li><li><a name="sequence.tests.test_sequence.TestValidSequence"></a><a href="sequence.tests.test_sequence.TestValidSequence.html">sequence.tests.test_sequence.TestValidSequence</a> - <span class="undocumented">No class docstring; 2/3 methods documented</span></li><li><a name="sequence.tests.test_transform.TestFastqClassMethods"></a><a href="sequence.tests.test_transform.TestFastqClassMethods.html">sequence.tests.test_transform.TestFastqClassMethods</a> - <span class="undocumented">No class docstring; 7/8 methods documented</span></li></ul></li></ul>
  </body>
</html>